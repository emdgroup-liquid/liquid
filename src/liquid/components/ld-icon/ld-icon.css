:host {
  display: inline-flex;

  > div {
    display: inline-flex;
  }
}
