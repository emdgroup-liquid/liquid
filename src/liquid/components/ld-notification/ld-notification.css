:host {
  /* layout */
  --ld-notification-min-height: var(--ld-sp-40);
  --ld-notification-border-radius: var(--ld-br-m);
  --ld-notification-inset-y: var(--ld-sp-32);

  /* colors */
  --ld-notification-col: var(--ld-col-wht);
  --ld-notification-warn-col: var(--ld-col-neutral-900);
  --ld-notification-alert-col: var(--ld-col-wht);

  /* themable colors */
  --ld-notification-bg-col: var(--ld-thm-primary);
  --ld-notification-warn-bg-col: var(--ld-thm-warning);
  --ld-notification-alert-bg-col: var(--ld-thm-error);

  position: fixed;
  z-index: 2147483647; /* Highest possible */
  display: flex;
  justify-content: center;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  min-width: 20rem;
  font: var(--ld-typo-body-m);
  pointer-events: none;
}

@keyframes ld-notification-fade-in {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}

@keyframes ld-notification-fade-out {
  from {
    transform: none;
    opacity: 1;
  }
  to {
    transform: none;
    opacity: 0;
  }
}

.ld-notification__item {
  pointer-events: auto;
  position: absolute;
  transform-origin: center;
  width: fit-content;
  min-width: 19rem;
  max-width: calc(90% - 1rem);
  max-height: 80vh;
  overflow-y: auto;
  display: flex;
  justify-content: space-between;
  transition: transform 0.2s ease, opacity 0.2s linear;
  box-shadow: var(--ld-shadow-sticky);
  min-height: var(--ld-notification-min-height);
  border-radius: var(--ld-notification-border-radius);

  &:where(:not(.ld-notification__item--dismissed)) {
    animation: ld-notification-fade-in 0.2s linear;
  }

  &:nth-last-of-type(n + 5) {
    opacity: 0;
  }

  &:nth-last-of-type(n + 2) {
    .ld-notification__item-content,
    .ld-notification__btn-dismiss {
      opacity: 0;
    }
  }

  &:nth-last-of-type(2) {
    opacity: 0.75;
  }

  &:nth-last-of-type(3) {
    opacity: 0.5;
  }

  &:nth-last-of-type(4) {
    opacity: 0.25;
  }

  &.ld-notification__item--dismissed {
    opacity: 0;
    animation: ld-notification-fade-out 0.2s linear forwards;
  }

  :host(.ld-notification--bottom) & {
    bottom: var(--ld-notification-inset-y);

    &:nth-last-of-type(2) {
      transform: scale(0.975) translateY(25%);
    }

    &:nth-last-of-type(3) {
      transform: scale(0.95) translateY(50%);
    }

    &:nth-last-of-type(4) {
      transform: scale(0.925) translateY(75%);
    }
  }

  :host(.ld-notification--top) & {
    top: var(--ld-notification-inset-y);

    &:nth-last-of-type(2) {
      transform: scale(0.975) translateY(-25%);
    }

    &:nth-last-of-type(3) {
      transform: scale(0.95) translateY(-50%);
    }

    &:nth-last-of-type(4) {
      transform: scale(0.925) translateY(-75%);
    }
  }
}

.ld-notification__item--info {
  background-color: var(--ld-notification-bg-col);
  color: var(--ld-notification-col);
}
.ld-notification__item--warn {
  background-color: var(--ld-notification-warn-bg-col);
  color: var(--ld-notification-warn-col);
}
.ld-notification__item--alert {
  background-color: var(--ld-notification-alert-bg-col);
  color: var(--ld-notification-alert-col);
}

.ld-notification__item-content,
.ld-notification__btn-dismiss {
  transition: opacity 0.2s linear;
}

.ld-notification__item-content {
  display: grid;
  grid-auto-flow: column;
  align-items: center;
  gap: var(--ld-sp-12);
  padding: var(--ld-sp-6) var(--ld-sp-12);

  > * {
    align-self: center;
  }
}

.ld-notification__btn-dismiss {
  margin-left: var(--ld-sp-16);
  align-self: center;
  flex-shrink: 0;
  border: 0;
  border-radius: var(--ld-br-full);
  background-color: transparent;
  cursor: pointer;
  padding: 0;
  height: var(--ld-sp-40);
  width: var(--ld-sp-40);
  display: inline-grid;
  place-items: center;
  color: inherit;
  user-select: none;
  touch-action: manipulation;
  -webkit-touch-callout: none;
}
