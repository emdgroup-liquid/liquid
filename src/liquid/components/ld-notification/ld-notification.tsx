import { Component, h, Host, Listen, Prop, State, Watch } from '@stencil/core'

type Notification = {
  type: 'info' | 'warn' | 'alert'
  content: string
  timeout?: number
}

const DEFAULT_NOTIFICATION_TIMEOUT = 6000
const FADE_TRANSITION_DURATION = 200

@Component({
  tag: 'ld-notification',
  styleUrl: 'ld-notification.shadow.css',
  shadow: true,
})
export class LdNotification {
  /**
   * Notification placement within the screen.
   */
  @Prop() placement?: 'top' | 'bottom' = 'top'

  @State() queue: Notification[] = []
  @State() queueDismissed: Notification[] = []

  @State() dismissTimeout: NodeJS.Timeout | null

  @State() fadeoutTimeouts: (NodeJS.Timeout | null)[] = []

  @State() currentNotification?: Notification

  @Watch('currentNotification')
  updateDismissTimeout() {
    clearTimeout(this.dismissTimeout)

    if (!this.currentNotification) return
    if (this.currentNotification.type === 'alert') return
    if (this.currentNotification.timeout === 0) return

    this.dismissTimeout = setTimeout(() => {
      this.handleNotificationDismiss()
    }, this.currentNotification.timeout || DEFAULT_NOTIFICATION_TIMEOUT)
  }

  @Listen('ldNotificationAdd', {
    target: 'window',
    passive: true,
  })
  handleNotification(ev: CustomEvent<Notification>) {
    ev.stopImmediatePropagation()
    const newNotification = ev.detail

    // If the same notification is already in queue (same content, same type), ignore this notification.
    const inQueue = this.queue.some(
      (notification) =>
        notification.content === newNotification.content &&
        notification.type === newNotification.type
    )
    if (inQueue) return

    // Insert by relevance, whith error notifications being more relevant than non-error notifications.
    if (newNotification.type === 'alert') {
      this.queue = [...this.queue, newNotification]
      this.currentNotification = newNotification
      return
    }

    const firstErrorNotificationIndex = this.queue.findIndex(
      (notification) => notification.type === 'alert'
    )
    if (firstErrorNotificationIndex === -1) {
      this.queue = [...this.queue, newNotification]
      this.currentNotification = newNotification
      return
    }

    this.queue.splice(firstErrorNotificationIndex, 0, newNotification)
    this.queue = [...this.queue]
  }

  @Listen('ldNotificationDismiss', {
    target: 'window',
    passive: true,
  })
  handleNotificationDismiss() {
    if (!this.currentNotification) return

    this.queueDismissed.unshift(this.queue.pop())
    this.queueDismissed = [...this.queueDismissed]
    this.queue = [...this.queue]
    this.currentNotification = this.queue[this.queue.length - 1]

    this.fadeoutTimeouts.push(
      setTimeout(() => {
        this.queueDismissed = this.queueDismissed.slice(0, -1)
      }, FADE_TRANSITION_DURATION)
    )
  }

  @Listen('ldNotificationClear', {
    target: 'window',
    passive: true,
  })
  handleNotificationClear() {
    this.queueDismissed = [...this.queue]
    this.queue = []
    this.currentNotification = undefined
    this.fadeoutTimeouts.forEach(clearTimeout)
    this.fadeoutTimeouts.push(
      setTimeout(() => {
        this.queueDismissed = []
      }, FADE_TRANSITION_DURATION)
    )
  }

  private renderNotification(notification: Notification, dismissed = false) {
    let cl = `ld-notification__item ld-notification__item--${notification.type}`
    if (dismissed) cl += ' ld-notification__item--dismissed'

    return (
      <div
        class={cl}
        key={notification.type + notification.content}
        part="item"
      >
        <div
          class="ld-notification__item-content"
          innerHTML={notification.content}
          role={notification.type === 'alert' ? 'alert' : 'status'}
          part="content"
        ></div>
        <button
          class="ld-notification__btn-dismiss"
          onClick={
            dismissed ? undefined : this.handleNotificationDismiss.bind(this)
          }
          part="btn-dismiss focusable"
        >
          {/* custom icon cross */}
          <svg
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            part="btn-dismiss-icon"
          >
            <title>Dismiss</title>
            <path
              d="M6 6L18 18"
              stroke="currentColor"
              stroke-width="3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
            <path
              d="M6 18L18 6"
              stroke="currentColor"
              stroke-width="3"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    )
  }

  disconnectedCallback() {
    clearTimeout(this.dismissTimeout)
    this.fadeoutTimeouts.forEach(clearTimeout)
  }

  render() {
    const cl = `ld-notification ld-notification--${this.placement}`

    return (
      <Host class={cl} role="region" aria-label="Notifications">
        {this.queue.map((notification) =>
          this.renderNotification.call(this, notification)
        )}
        {this.queueDismissed.map((notification) =>
          this.renderNotification.call(this, notification, true)
        )}
      </Host>
    )
  }
}
