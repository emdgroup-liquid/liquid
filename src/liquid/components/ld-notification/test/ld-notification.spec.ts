import { newSpecPage } from "@stencil/core/testing";
import { LdNotification } from "../ld-notification";
import { LdIcon } from "../../ld-icon/ld-icon";

const DEFAULT_NOTIFICATION_TIMEOUT = 6000;
const FADE_TRANSITION_DURATION = 200;

describe("ld-notification", () => {
  afterEach(() => {
    jest.clearAllTimers();
  });

  it("renders as empty container placed at the top by default", async () => {
    const page = await newSpecPage({
      components: [LdNotification],
      html: `<ld-notification></ld-notification>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  describe("placement", () => {
    it('renders placed at the top with prop placement set to "top"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification placement="top"></ld-notification>`,
      });
      const ldNotification = page.root;
      expect(
        ldNotification.classList.contains("ld-notification--bottom"),
      ).toBeFalsy();
      expect(
        ldNotification.classList.contains("ld-notification--top"),
      ).toBeTruthy();
    });

    it('renders placed at the bottom with prop placement set to "bottom"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification placement="bottom"></ld-notification>`,
      });
      const ldNotification = page.root;
      expect(
        ldNotification.classList.contains("ld-notification--bottom"),
      ).toBeTruthy();
      expect(
        ldNotification.classList.contains("ld-notification--top"),
      ).toBeFalsy();
    });
  });

  describe("notification types", () => {
    it('renders a notification of type "info"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item",
      );

      expect(notifications.length).toEqual(1);

      const notification = notifications[0];
      expect(notification.textContent).toContain("I am an info message.");

      expect(
        notification.classList.contains("ld-notification__item--info"),
      ).toBeTruthy();
      expect(
        notification.classList.contains("ld-notification__item--warn"),
      ).toBeFalsy();
      expect(
        notification.classList.contains("ld-notification__item--alert"),
      ).toBeFalsy();
    });

    it('renders a notification of type "info" if type is omitted', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
          },
        }),
      );
      await page.waitForChanges();

      const notification = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item",
      )[0];
      expect(
        notification.classList.contains("ld-notification__item--info"),
      ).toBeTruthy();
    });

    it('renders a notification of type "warn"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am a warning.",
            type: "warn",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item",
      );

      expect(notifications.length).toEqual(1);

      const notification = notifications[0];
      expect(notification.textContent).toContain("I am a warning.");

      expect(
        notification.classList.contains("ld-notification__item--info"),
      ).toBeFalsy();
      expect(
        notification.classList.contains("ld-notification__item--warn"),
      ).toBeTruthy();
      expect(
        notification.classList.contains("ld-notification__item--alert"),
      ).toBeFalsy();
    });

    it('renders a notification of type "alert"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "Ooops.",
            type: "alert",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item",
      );

      expect(notifications.length).toEqual(1);

      const notification = notifications[0];
      expect(notification.textContent).toContain("Ooops.");

      expect(
        notification.classList.contains("ld-notification__item--info"),
      ).toBeFalsy();
      expect(
        notification.classList.contains("ld-notification__item--warn"),
      ).toBeFalsy();
      expect(
        notification.classList.contains("ld-notification__item--alert"),
      ).toBeTruthy();
    });
  });

  describe("notification hierarchy", () => {
    it('queues potentially less important notifications behind notifications of type "alert"', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "warning #1",
            type: "warn",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "alert #1",
            type: "alert",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #2",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "warning #2",
            type: "warn",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "alert #2",
            type: "alert",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #3",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item",
      );

      expect(notifications.length).toEqual(7);

      expect(notifications[6].textContent).toContain("alert #2");
      expect(notifications[5].textContent).toContain("alert #1");
      expect(notifications[4].textContent).toContain("info #3");
      expect(notifications[3].textContent).toContain("warning #2");
      expect(notifications[2].textContent).toContain("info #2");
      expect(notifications[1].textContent).toContain("warning #1");
      expect(notifications[0].textContent).toContain("info #1");
    });
  });

  describe("notification timeout", () => {
    it("dismisses current notification on notification dismiss button click", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      const btnDismiss = notifications[0].querySelector(
        ".ld-notification__btn-dismiss",
      );
      expect(btnDismiss).toBeTruthy();

      btnDismiss.dispatchEvent(new Event("click"));
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);
    });

    it('dismisses a notifications of type "info" after default timeout', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);
    });

    it('dismisses a notifications of type "info" after custom timeout', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      const additionalTime = 200;
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
            timeout: DEFAULT_NOTIFICATION_TIMEOUT + additionalTime,
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(additionalTime);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);
    });

    it('does not dismiss a notification of type "info" with a timeout set to 0', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
            timeout: 0,
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);
    });

    it('does not dismiss a notification of type "alert" after default timeout', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "Ooops.",
            type: "alert",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);
    });

    it('dismisses a notification of type "alert" after explicit timeout', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "Ooops.",
            type: "alert",
            timeout: DEFAULT_NOTIFICATION_TIMEOUT,
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);
    });

    it('does not dismiss a notification of type "info" if it is in queue behind another notification', async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "I am an info message.",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "Ooops.",
            type: "alert",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(DEFAULT_NOTIFICATION_TIMEOUT);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);
    });
  });

  describe("dismissing notifications", () => {
    it("dismisses current notification on ldNotificationDismiss event", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #2",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      page.win.dispatchEvent(new CustomEvent("ldNotificationDismiss"));
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);
      expect(notifications[0].textContent).toContain("info #1");
    });

    it("does not do anything on ldNotificationDismiss event if no notifications are queued", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      await page.waitForChanges();

      page.win.dispatchEvent(new CustomEvent("ldNotificationDismiss"));
      await page.waitForChanges();

      expect(page.root).toEqualHtml(`
        <ld-notification aria-label="Notifications" class="ld-notification ld-notification--top" role="region">
          <mock:shadow-root></mock:shadow-root>
        </ld-notification>
      `);
    });

    it("dismisses all notifications on ldNotificationClear event", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #2",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      page.win.dispatchEvent(new CustomEvent("ldNotificationClear"));
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);
    });
  });

  describe("transitions", () => {
    it("transitions single notifications", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #2",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #3",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #4",
            type: "info",
          },
        }),
      );

      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(4);

      let dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(0);

      page.win.dispatchEvent(new CustomEvent("ldNotificationDismiss"));
      await page.waitForChanges();

      page.win.dispatchEvent(new CustomEvent("ldNotificationDismiss"));
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(2);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(FADE_TRANSITION_DURATION);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(0);
    });

    it("transitions multiple notifications", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #2",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      let notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(2);

      let dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(0);

      page.win.dispatchEvent(new CustomEvent("ldNotificationClear"));
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);

      dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(2);

      // Fast-forward until the timer has been executed.
      jest.advanceTimersByTime(FADE_TRANSITION_DURATION);
      await page.waitForChanges();

      notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(0);

      dismissedNotifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item--dismissed",
      );
      expect(dismissedNotifications.length).toEqual(0);
    });
  });

  describe("notification content", () => {
    it("ignores redundant notifications", async () => {
      const page = await newSpecPage({
        components: [LdNotification],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );

      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: "info #1",
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);
    });

    it("renders notifications with HTML content", async () => {
      const page = await newSpecPage({
        components: [LdNotification, LdIcon],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content:
              '<ld-icon name="placeholder"></ld-icon> A notification with an icon.',
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      const svg = notifications[0].querySelector("svg");
      expect(svg).toBeTruthy();
    });

    it("it sanitizes notifications with HTML content", async () => {
      const page = await newSpecPage({
        components: [LdNotification, LdIcon],
        html: `<ld-notification></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: '<img alt=1 src=1 href=1 onerror="javascript:alert(1)">',
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      const img = notifications[0].querySelector("img");
      expect(img).toBeTruthy();

      expect(img.hasAttribute("alt")).toBeTruthy();
      expect(img.hasAttribute("src")).toBeTruthy();
      expect(img.hasAttribute("href")).toBeTruthy();
      expect(img.hasAttribute("onerror")).toBeFalsy();
    });

    it("it sanitizes notifications with HTML content and custom sanitization options", async () => {
      const page = await newSpecPage({
        components: [LdNotification, LdIcon],
        html: `<ld-notification sanitize-config='{"FORBID_ATTR": ["href"]}'></ld-notification>`,
      });
      page.win.dispatchEvent(
        new CustomEvent("ldNotificationAdd", {
          detail: {
            content: '<img alt=1 src=1 href=1 onerror="javascript:alert(1)">',
            type: "info",
          },
        }),
      );
      await page.waitForChanges();

      const notifications = page.root.shadowRoot.querySelectorAll(
        ".ld-notification__item:not(.ld-notification__item--dismissed)",
      );
      expect(notifications.length).toEqual(1);

      const img = notifications[0].querySelector("img");
      expect(img).toBeTruthy();

      expect(img.hasAttribute("alt")).toBeTruthy();
      expect(img.hasAttribute("src")).toBeTruthy();
      expect(img.hasAttribute("href")).toBeFalsy();
      expect(img.hasAttribute("onerror")).toBeFalsy();
    });
  });

  it("does not throw when disconnecting before hydration", () => {
    const component = new LdNotification();
    component.disconnectedCallback();
  });
});
