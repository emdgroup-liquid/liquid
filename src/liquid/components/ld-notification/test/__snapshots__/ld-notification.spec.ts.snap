// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-notification renders as empty container placed at the top by default 1`] = `
<ld-notification aria-label="Notifications" class="ld-notification ld-notification--top" role="region">
  <mock:shadow-root></mock:shadow-root>
</ld-notification>
`;
